minimal-bootstrap: make sources a non-tarballs.nixos.org FOD

amjoseph-nixpkgs opened the pull request:
Alternative to https://github.com/NixOS/nixpkgs/pull/232576, does not require manual uploads to `tarballs.nixos.org`.

###### Description of changes

This commit adjusts #232576 to break the fetchurl<->minimal-bootstrap-sources dependency cycle without needing an upload to tarballs.nixos.org.  It does this by appending the low-level FOD attributes onto the `runCommand` derivation.

  https://nixos.org/manual/nix/unstable/language/advanced-attributes.html#adv-attr-outputHash

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

emilytrau committed:
minimal-bootstrap: create top-level attr for bootstrap sources

Commit:
minimal-bootstrap: make sources a non-tarballs.nixos.org FOD

This commit adjusts #232576 to break the
fetchurl<->minimal-bootstrap-sources dependency cycle without
needing an upload to tarballs.nixos.org.  It does this by appending
the low-level FOD attributes onto the `runCommand` derivation.

  https://nixos.org/manual/nix/unstable/language/advanced-attributes.html#adv-attr-outputHash

amjoseph-nixpkgs commented:
The gigantic comment should be moved into the manual once we switch over to using `minimal-bootstrap` as our source of `bootstrapFiles`.

emilytrau commented on pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix:
```diff
@@ -15,20 +18,63 @@ rec {
     executable = true;
```
> While it is included in the stage0-posix source bundle and is synced with

Would you mind updating this comment while you're at it to reflect that this binary is no longer in the source bundle?

amjoseph-nixpkgs replied:
908b2a51430e

emilytrau commented on pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/make-bootstrap-sources.nix:
```diff
@@ -9,23 +8,30 @@
 # To build:
 #
-#   nix-build pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/make-bootstrap-sources.nix
+#   nix-build '<nixpkgs>' -A minimal-bootstrap-sources
```
Please change this to the new way of updating the narball hash :)

amjoseph-nixpkgs replied:
a877fb253db2

emilytrau commented on pkgs/os-specific/linux/minimal-bootstrap/default.nix:
```diff
@@ -4,6 +4,7 @@
+, make-minimal-bootstrap-sources
 }:
 
 lib.makeScope
```
Add `make-minimal-bootstrap-sources` to the scope inputs here so it doesn't need to be manually passed through?

amjoseph-nixpkgs replied:
5dacb6894093

emilytrau replied:
Sorry should clarify that this line would be the best place to declare it :) https://github.com/NixOS/nixpkgs/blob/3b5e8694a52cdcc07a32f324ebe5f0ddbd14fe33/pkgs/os-specific/linux/minimal-bootstrap/default.nix#L14

amjoseph-nixpkgs replied:
3d3fe75cbc6f

emilytrau reviewed (changes requested):
This is awesome! Super creative way of solving the bootstrap sources issue! 😍 

Have you got a method to validate that this won't have infinite recursion once we integrate into the `bootstrap-tools` dependency chain? I'm not sure how to test that

emilytrau commented:
@amjoseph-nixpkgs would you be interested in adding yourself the `minimal-bootstrap` maintainers team as well?

amjoseph-nixpkgs commented:
Squashed.

> Have you got a method to validate that this won't have infinite recursion once we integrate into the `bootstrap-tools` dependency chain? I'm not sure how to test that

Yes, we can have a test which initializes a completely empty store and does a `nix-instantiate`, like [this](https://github.com/NixOS/nixpkgs/blob/c15d559895812d5cdcbb707215ac75072d56f7fd/lib/tests/release.nix#L44)

> @amjoseph-nixpkgs would you be interested in adding yourself the `minimal-bootstrap` maintainers team as well?

Sure; I probably need to add myself to `maintainer-list.nix` first though.  Procrastination, procrastination...

emilytrau commented on pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/default.nix:
```diff
@@ -1,9 +1,13 @@
 { lib
 , newScope
+, make-minimal-bootstrap-sources
```
This is in scope and not needed to be passed through

amjoseph-nixpkgs replied:
52edaab39be9

emilytrau reviewed (changes requested):
One last thing and very happy to go! 🚀🚀

Ericson2314 commented on pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/make-bootstrap-sources.nix:
```diff
@@ -35,12 +40,20 @@ let
+
 in
-runCommand name {
+(runCommand "${name}.nar.xz" {
```
Paren shouldn't be needed?

amjoseph-nixpkgs replied:
1f87f1c02a94

Ericson2314 commented on pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix:
```diff
@@ -1,34 +1,77 @@
+  specifically designed for bit-exact reproducibility, none of the
+  requirements above apply to `minimal-bootstrap-sources`.
+  */
+  minimal-bootstrap-sources = make-minimal-bootstrap-sources.overrideAttrs(_: {
```
This will eventually not work because the drv files are cyclic (cyclic hashing problem) even if the input-addressing isn't cyclic. But that's fine. We can use a stub failing FOD with the same hash at that point,

amjoseph-nixpkgs replied:
You're absolutely right.  We should just do it now and get it over with.  @Ericson2314 @emilytrau, please take a look at eb2e0bb9421e which implements this.  Here's what the failure looks like on the console:

![bootfail](https://github.com/NixOS/nixpkgs/assets/54836058/314916e6-568d-4fe8-87b6-ddf62ef255d3)

amjoseph-nixpkgs replied:
I tested this by making `bootstrapTools` depend on `minimal-bootstrap-tools` as shown below, and eval succeeded.  The build began, but obviously it won't complete.

```diff
diff --git a/pkgs/stdenv/linux/bootstrap-tools/default.nix b/pkgs/stdenv/linux/bootstrap-tools/default.nix
index 569f0c6f31e2..6aa8e6fbc6cf 100644
--- a/pkgs/stdenv/linux/bootstrap-tools/default.nix
+++ b/pkgs/stdenv/linux/bootstrap-tools/default.nix
@@ -3,7 +3,7 @@
 derivation ({
   name = "bootstrap-tools";

-  builder = bootstrapFiles.busybox;
+  builder = "/bin/sh";

   args = [ "ash" "-e" ./scripts/unpack-bootstrap-tools.sh ];

diff --git a/pkgs/stdenv/linux/default.nix b/pkgs/stdenv/linux/default.nix
index 9cfe21e3640d..e475787287d6 100644
--- a/pkgs/stdenv/linux/default.nix
+++ b/pkgs/stdenv/linux/default.nix
@@ -55,6 +55,8 @@
 #     $ nix-tree --derivation $(nix-instantiate -A stdenv)
 { lib
 , localSystem, crossSystem, config, overlays, crossOverlays ? []
+, minimal-bootstrap-tools ?
+  (import ../../os-specific/linux/minimal-bootstrap/stage0-posix/bootstrap-sources.nix {}).minimal-bootstrap-sources

 , bootstrapFiles ?
   let table = {
@@ -92,7 +94,10 @@
     or (abort "unsupported libc for the pure Linux stdenv");
   files = archLookupTable.${localSystem.system} or (if getCompatibleTools != null then getCompatibleTools
     else (abort "unsupported platform for the pure Linux stdenv"));
-  in files
+  in {
+    bootstrapTools = minimal-bootstrap-tools;
+    busybox = null;
+  }
 }:

 assert crossSystem == localSystem;
```

Ericson2314 commented on pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/make-bootstrap-sources.nix:
```diff
@@ -35,12 +40,20 @@ let
-  nix-store --dump ${src} | xz -c > "$out/${name}.nar.xz"
-''
+  nix-store --dump ${src} | xz -c > $out
+'')
```
And likewise the closing paren

amjoseph-nixpkgs replied:
1f87f1c02a94

Ericson2314 commented on pkgs/os-specific/linux/minimal-bootstrap/stage0-posix/default.nix:
```diff
@@ -1,9 +1,13 @@
 
 lib.makeScope newScope (self: with self; {
-  inherit (import ./bootstrap-sources.nix) version hex0-seed src;
+  inherit (import ./bootstrap-sources.nix { inherit make-minimal-bootstrap-sources; })
```
I suppose if we remove `make-minimal-bootstrap-sources` from above as @emilytrau suggests then we need to use `self.callPackage` here, right? Sounds good to me.

amjoseph-nixpkgs replied:
52edaab39be9

emilytrau commented:
Linking back to the thread #227914

Commit:
separate minimal-bootstrap-sources from make-minimal-bootstrap-sources

Ericson2314 observed that although FODs break the build-time cycle,
they don't break the eval-time cycle.  This cycle must be broken in
order to compute the derivation hash.

make-minimal-bootstrap-sources does and always has depended on the
full nixpkgs (including fetchFromGitHub).  This commit completely
separates minimal-bootstrap-sources from it, so that
minimal-bootstrap-sources now does not depend on
make-minimal-bootstrap-sources (or nixpkgs) in any way.

minimal-bootstrap-sources is now a "bare Nix" derivation.  It is an
FOD whose builder always fails with an error message.

Ericson2314 reviewed (approved):
Looks good to me!

Ericson2314 commented:
Uh oh, did I merge to soon? the hex1 build seems to be spinning in a loop for me.
